llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,504 @@
+  std::error_code EC;
+  sys::fs::mapped_file_region Map(FD, sys::fs::mapped_file_region::readonly,
+                                  FS->getSize(),
+                                  /*offset=*/0, EC);
+  if (EC)
+    return errorCodeToError(EC);
+
+  auto &SaveQueue = CASDB->saveQueue();
+  ArrayRef<char> Data(Map.data(), Map.size());
+  SmallString<128> Path;
+  if (sys::fs::getRealPathFromHandle(FD, Path))
+    SaveQueue.saveFileAsync(Path.str().str());
+  else
+    SaveQueue.saveDataAsync(toStringRef(Data).str());
+
+  auto Response = SaveQueue.receiveNext();
+  if (!Response)
+    return Response.takeError();
+
+  auto &I = indexHash(arrayRefFromStringRef(Response->CASID));
+  // TODO: we can avoid the copy by implementing InMemoryRef object like
+  // InMemoryCAS.
+  return toReference(storeObjectImpl(I, None, Data));
+}
+
+Expected<ObjectRef> GRPCRelayCAS::storeFile(StringRef Content) {
+  auto &SaveQueue = CASDB->saveQueue();
+  SaveQueue.saveDataAsync(Content.str());
+  auto Response = SaveQueue.receiveNext();
+  if (!Response)
+    return Response.takeError();
+
+  auto &I = indexHash(arrayRefFromStringRef(Response->CASID));
+  ArrayRef<char> Data = arrayRefFromStringRef<char>(Content);
+  return toReference(storeObjectImpl(I, {}, Data));
+}
+
+Expected<StringRef> GRPCRelayCAS::loadFile(ObjectRef Ref) {
+  auto &I = asInMemoryIndexValue(Ref);
+
+  // Return the existing value if we have one.
+  if (const InMemoryCASData *Existing = I.Data.load()) {
+    assert(Existing->getNumRefs() == 0 && "File should have zero refs");
+    return toStringRef(Existing->getData());
+  }
+
+  auto &LoadQueue = CASDB->loadQueue();
+  LoadQueue.loadAsync(getDataIDFromRef(Ref));
+  auto Response = LoadQueue.receiveNext();
+
+  auto getDataAsStringRef =
+      [&](Expected<ObjectHandle> Handle) -> Expected<StringRef> {
+    if (!Handle)
+      return Handle.takeError();
+    return toStringRef(getData(*Handle));
+  };
+  assert(Response->BlobData && "Expect the result to be returned in Blob");
+  std::string Data = *Response->BlobData;
+  ArrayRef<char> Content(Data.data(), Data.size());
+  return getDataAsStringRef(getObjectHandle(storeObjectImpl(I, {}, Content)));
+}
+
+GRPCActionCache::GRPCActionCache(StringRef Path, Error &Err)
+    : ActionCache(GRPCCASContext::getDefaultContext()) {
+  ErrorAsOutParameter ErrAsOutParam(&Err);
+  auto Cache = createRemoteKeyValueClient(Path);
+  if (!Cache) {
+    Err = Cache.takeError();
+    return;
+  }
+  KVDB = std::move(*Cache);
+}
+
+Expected<Optional<CASID>>
+GRPCActionCache::getImpl(ArrayRef<uint8_t> ResolvedKey) const {
+  auto &GetValueQueue = KVDB->getValueQueue();
+  GetValueQueue.getValueAsync(ResolvedKey);
+  auto Response = GetValueQueue.receiveNext();
+  if (!Response)
+    return Response.takeError();
+  if (!Response->Value)
+    return None;
+
+  auto Result = Response->Value->find("CASID");
+  if (Result == Response->Value->end())
+    return None;
```

Comment:
This seems like it should be an error, because it would be unexpected to get a result that didn't include "CASID", something went wrong.